File: aacalc.pl
#!/usr/bin/perl
use Storable qw(dclone);
use Memoize;

memoize('binPDF');

#I don't think memoizing these will speed things up
#memoize('factorial'); memoize('power');
#memoize('min'); memoize('max');

#Memoizing doRound saves a *huge* amount of work.  In one test of
#10 Inf vs 10 Inf the processing time went from 70+ minutes (I stopped it)
#to a mere 1.4 seconds.  And the difference gets more impressive with
#more troops.

memoize('doRound');

@possible_units = ( 'Infantry', 'Artillary', 'Tanks', 'Fighters', 'Bombers' );

#------------------------------
$VERBOSE  = 0;
$VERBOSE2 = 0;

if (@ARGV) {
    %troops = long_units(@ARGV);
}
else {
    $troops{'Att'}{'Infantry'}  = 10;
    $troops{'Att'}{'Artillary'} = 0;
    $troops{'Att'}{'Tanks'}     = 3;
    $troops{'Att'}{'Fighters'}  = 0;
    $troops{'Att'}{'Bombers'}   = 0;

    $troops{'Def'}{'Infantry'}  = 10;
    $troops{'Def'}{'Artillary'} = 0;
    $troops{'Def'}{'Tanks'}     = 0;
    $troops{'Def'}{'Fighters'}  = 0;
    $troops{'Def'}{'Bombers'}   = 0;
}

#------------------------------

$odds{'Infantry'}{'Att'}  = 1;
$odds{'Infantry'}{'Def'}  = 2;
$odds{'Artillary'}{'Att'} = 2;
$odds{'Artillary'}{'Def'} = 2;
$odds{'Tanks'}{'Att'}     = 3;
$odds{'Tanks'}{'Def'}     = 2;
$odds{'Fighters'}{'Att'}  = 3;
$odds{'Fighters'}{'Def'}  = 4;
$odds{'Bombers'}{'Att'}   = 4;
$odds{'Bombers'}{'Def'}   = 1;

### Main ##############################################################
print STDERR ( ( join ',', 'Outcome', 'Odds', short_units_hdr() ), "\n" )
  if $VERBOSE;

# Here we do the "promotion" of troops in the presense of artillary
# Since these troops attack as artillary, we just make them artillary
# and keep track of things so we can reduce the number at the end (if
# any are left)
my $orig_infantry = $troops{'Att'}{'Infantry'};    #save this

my $promotions = min( $troops{'Att'}{'Infantry'}, $troops{'Att'}{'Artillary'} );
$troops{'Att'}{'Infantry'} -= $promotions;
$troops{'Att'}{'Artillary'} += $promotions;

( $win_pct, $lose_pct, $tie_pct ) = doRound( short_units(%troops) );

print( 'WIN,LOSE,TIE', "\n" );
my $total = ( $win_pct + $lose_pct + $tie_pct );
print(
    join ",",
    (
        $win_pct / $total * 100,
        $lose_pct / $total * 100,
        $tie_pct / $total * 100
    ),
    "\n"
);

#print ((join ",", ($win_pct,$lose_pct,$tie_pct)), "\n");
### End Main ##########################################################
exit int( $win_pct / ( $win_pct + $lose_pct ) * 100 );

sub doRound {
    print "Round\t", $round++, "\n" if ( $VERBOSE and not( $round % 10 ) );
    print "Called doRound(@_)", "\n" if $VERBOSE2;

    #process a round
    my %units = long_units(@_);

    #figure out if someone is dead
    my $Att_units_alive = 0;
    my $Def_units_alive = 0;

    foreach $unittype (@possible_units) {
        $Att_units_alive += $units{'Att'}{$unittype};
        last if $Att_units_alive;
    }
    foreach $unittype (@possible_units) {
        $Def_units_alive += $units{'Def'}{$unittype};
        last if $Def_units_alive;
    }

    if ( $Att_units_alive == 0 and $Def_units_alive == 0 ) {

        #TIE
        print STDERR ( ( join ',', 'TIE', short_units(%units) ), "\n" )
          if $VERBOSE;
        return ( 0, 0, 1 );
    }
    elsif ( $Att_units_alive == 0 ) {

        #LOSE
        print STDERR ( ( join ',', 'LOSE', short_units(%units) ), "\n" )
          if $VERBOSE;
        return ( 0, 1, 0 );
    }
    elsif ( $Def_units_alive == 0 ) {

        #WIN
        print STDERR ( ( join ',', 'WIN', short_units(%units) ), "\n" )
          if $VERBOSE;
        return ( 1, 0, 0 );
    }

    my %max_hits;
    my %hits_prob;
    my @hits;
    foreach $side ( 'Att', 'Def' ) {
        $max_hits{$side} = sum( values %{ $units{$side} } );
        foreach $infantry_hits ( 0 .. $units{$side}{'Infantry'} ) {
            foreach $artillary_hits ( 0 .. $units{$side}{'Artillary'} ) {
                foreach $tank_hits ( 0 .. $units{$side}{'Tanks'} ) {
                    foreach $fighter_hits ( 0 .. $units{$side}{'Fighters'} ) {
                        foreach $bomber_hits ( 0 .. $units{$side}{'Bombers'} ) {

                            $hits_prob{$side}[ $infantry_hits +
                              $artillary_hits +
                              $tank_hits +
                              $fighter_hits +
                              $bomber_hits ] += binPDF(
                                $odds{'Infantry'}{$side} / 6,
                                $units{$side}{'Infantry'},
                                $infantry_hits
                              ) * binPDF(
                                $odds{'Artillary'}{$side} / 6,
                                $units{$side}{'Artillary'},
                                $artillary_hits
                              ) * binPDF( $odds{'Tanks'}{$side} / 6,
                                $units{$side}{'Tanks'}, $tank_hits ) * binPDF(
                                $odds{'Fighters'}{$side} / 6,
                                $units{$side}{'Fighters'},
                                $fighter_hits
                                ) * binPDF( $odds{'Bombers'}{$side} / 6,
                                $units{$side}{'Bombers'}, $bomber_hits );
                        }
                    }
                }
            }
        }
    }
    my ( $prob_win, $prob_lose, $prob_tie ) = ( 0, 0, 0 );
    my ( $ahp, $dhp ) = ( 0, 0 );
    my $odds = 0;
    foreach $ahp ( 0 .. $max_hits{'Att'} ) {
        foreach $dhp ( 0 .. $max_hits{'Def'} ) {
            next if ( $ahp == 0 and $dhp == 0 );
            $odds = $hits_prob{'Att'}[$ahp] * $hits_prob{'Def'}[$dhp];
            ( $return_win, $return_lose, $return_tie ) =
              doRound(
                short_units( casualties( $ahp, $dhp, dclone( \%units ) ) ) );

 #multiply the probabilities of win and loss by the probability of us being here
            $prob_win  += $return_win * $odds;
            $prob_lose += $return_lose * $odds;
            $prob_tie  += $return_tie * $odds;
        }
    }

   #return ($prob_win/($prob_win+$prob_lose),$prob_lose/($prob_win+$prob_lose));
    return ( $prob_win, $prob_lose, $prob_tie );
}

sub casualties {

    #Decide on which units to remove
    #($hits{'Att'},$hits{'Def'},%cas_units) = @_;
    my %hits;

    # Reverse Attack and defence hits.  Before it meant "how many hits did
    # that side get."  Now it means "how many times did that side get hit."
    ( $hits{'Def'} ) = shift @_;
    ( $hits{'Att'} ) = shift @_;
    @orig_units = (@_);
    my %cas_units;
    my $reduce;

    foreach $side ( 'Att', 'Def' ) {
        foreach $trooptype (@possible_units) {
            $reduce =
              min( $hits{$side}, $orig_units[0]->{$side}{$trooptype} ) + 0;
            $cas_units{$side}{$trooptype} =
              $orig_units[0]->{$side}{$trooptype} - $reduce;
            $hits{$side} -= $reduce;
        }
    }
    return %cas_units;
}

sub max {
    return $_[0] if $_[0] >= $_[1];
    return $_[1];
}

sub min {
    return $_[0] if $_[0] <= $_[1];
    return $_[1];
}

sub binPDF {
    my ( $P, $N, $K ) = @_;

    # $P = probability of hitting
    # $N = number of dice thrown
    # $K = number of kills
    # returns chance of that many hits happening

    return 1 if ( $N == 0 );
    return power( 1 - $P, $N ) if ( $K == 0 );
    return factorial($N) /
      ( factorial($K) * factorial( $N - $K ) ) *
      power( $P,     $K ) *
      power( 1 - $P, $N - $K );
}

sub binPDF2 {
    my ( $P, $N, $K ) = @_;
    print "$P,$N,$K\n";
    return 1 if ( $N == 0 );
    return power( 1 - $P, $N ) if ( $K == 0 );
    return factorial($N) /
      ( factorial($K) * factorial( $N - $K ) ) *
      power( $P,     $K ) *
      power( 1 - $P, $N - $K );

}

sub factorial {
    my $n = shift;
    return undef if $n < 0;
    return 1 if $n == 0;
    return $n * factorial( $n - 1 );
}

sub power {
    return $_[0] ** $_[1];
}

sub sum {
    my (@array) = @_;
    my $sum;
    $sum = 0;
    while (@array) { $sum += pop(@array) }
    return $sum;
}

sub short_units {
    my %unit_hash = @_;
    my @return_array;

    foreach $side ( 'Att', 'Def' ) {
        foreach $unit (@possible_units) {
            @return_array = ( @return_array, $unit_hash{$side}{$unit} );
        }
    }
    return @return_array;
}

sub short_units_hdr {
    my @return_array;
    foreach $side ( 'Att', 'Def' ) {
        foreach $unit (@possible_units) {
            @return_array = ( @return_array, $side . $unit );
        }
    }
    return @return_array;
}

sub long_units {
    my @short_units = @_;
    my %return_hash;

    foreach $side ( 'Att', 'Def' ) {
        foreach $unit (@possible_units) {
            $return_hash{$side}{$unit} = ( shift @short_units ) + 0;
        }
    }
    return %return_hash;
}

sub print_results {
    my @p_results   = (@_);
    my $num_records = ( 1 + 1 + ( @possible_units * 2 ) );
    my $num_rows    = ( @p_results + 0 ) / $num_records;

    foreach $x ( 1 .. $num_rows ) {
        my @curr_row = ();
        foreach $y ( 1 .. $num_records ) {
            @curr_row = ( @curr_row, shift @p_results );
        }
        print join (",", @curr_row ), "\n";
    }
}

# vi:ts=4:

File: aacalc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set tabstop=4 shiftwidth=4 autoindent expandtab :
# pylint: disable=line-too-long
# pylint: disable=invalid-name
"""
aacalc.py - A script for calculating the odds of winning a battle in a game.

This script contains several functions for calculating the odds of winning a battle in a game. The script imports the necessary modules and defines some global variables, including the `POSSIBLE_UNITS` list and the `odds` dictionary.

The script can be run from the command line with the appropriate arguments to calculate the odds of winning a battle. The script also includes some Vim configuration comments and pylint configuration comments.

Functions:
    binPDF(p, n, k): Calculates the binomial probability mass function.
    casualties(att_hits, def_hits, units): Applies the hits to the units and returns the new units.

Global Constants:
    possible_units: A list of the possible unit types in the game.
    odds: A dictionary containing the odds of hitting for each unit type on each side.

Usage:
    python aacalc.py [att_inf att_art att_tan att_fig att_bom def_inf def_art def_tan def_fig def_bom]

Example:
    python aacalc.py 10 5 3 2 1 8 3 2 1 0
Output:
    99.90,0.08,0.01
"""

from math import (comb, prod)
from copy import deepcopy
from itertools import product
from functools import lru_cache

# Global constants
VERBOSE = 0
POSSIBLE_UNITS = ['Infantry', 'Artillery', 'Tanks', 'Fighters', 'Bombers']
ODDS = {
    'Infantry': {'Att': 1, 'Def': 2},
    'Artillery': {'Att': 2, 'Def': 2},
    'Tanks': {'Att': 3, 'Def': 2},
    'Fighters': {'Att': 3, 'Def': 4},
    'Bombers': {'Att': 4, 'Def': 1}
}

class Units:
    """
    The Units class is used to store the number of units of each type on each side of the battle.
    The class also contains a method for applying the hits to the units and returning the new units.
    If called as a hash return the frozenset of the units.
    """
    def __init__(self, inUnits=None):
        if inUnits is None:
            self.att = {}
            self.def_ = {}
        elif isinstance(inUnits, list) or isinstance(inUnits, tuple):
            self.att = {unit: inUnits[i] for i, unit in enumerate(POSSIBLE_UNITS)}
            self.def_ = {unit: inUnits[i + 5] for i, unit in enumerate(POSSIBLE_UNITS)}
        elif isinstance(inUnits, dict):
            self.att = inUnits['Att']
            self.def_ = inUnits['Def']
        elif isinstance(inUnits, Units):
            self.att = deepcopy(inUnits.att)
            self.def_ = deepcopy(inUnits.def_)
        else:
            raise ValueError('Invalid input type')
        # Set all units to 0 if they are not present
        for side in [self.att, self.def_]:
            for unit in POSSIBLE_UNITS:
                if unit not in side:
                    side[unit] = 0

    def __list__(self):
        return list(self.att.values()) + list(self.def_.values())

    def __str__(self):
        return str(self.__list__())

    def __hash__(self):
        return_code = self.__str__()
        return hash(return_code)

    def __dict__(self):
        return {'Att': self.att, 'Def': self.def_}

    def values(self):
        """Convert to list"""
        return self.__list__()

    def units(self, side):
        """Units by side"""
        return self.att if side == 'Att' else self.def_

    def assign_hits(self, side, unittype, hits):
        """Assign hits to units"""
        if side == 'Att':
            self.att[unittype] -= hits
        else:
            self.def_[unittype] -= hits

    def alive(self,side):
        """Returns the number of units alive on the given side"""
        return sum(self.att.values()) if side == 'Att' else sum(self.def_.values())

@lru_cache(maxsize=None)
def binPDF(P: float, N: int, K: int) -> float:
    """
    Binomial Probability Distribution Function
    P = probability of hitting
    N = number of dice thrown
    K = number of kills
    returns chance of that many hits happening (0,1)
    """
    if N == 0:
        return 1.0
    if K == 0:
        return (1 - P) ** N
    return comb(N, K) * P ** K * (1 - P) ** (N - K)

def casualties(att_hits, def_hits, orig_units):
    """
    Applies the hits to the units and returns the new units.
    """
    units = Units(orig_units)
    # Reverse Attack and defence hits.  Before it meant "how many hits did
    # that side get."  Now it means "how many times did that side get hit."
    hits = {'Att': def_hits, 'Def': att_hits}
    for side in ['Att', 'Def']:
        for unittype in POSSIBLE_UNITS:
            reduce = min(hits[side], units.units(side)[unittype])
            units.assign_hits(side, unittype, reduce)
            hits[side] -= reduce
    return units

# Can take either 1 argument of type Units, or 10 items
@lru_cache(maxsize=None)
def doRound(*args):
    """
    Calculates a round of combat.
    """
    if len(args) == 1 and isinstance(args[0], Units):
        units = args[0]
    elif len(args) == 10:
        units = Units(args)
    else:
        raise ValueError("doRound expects either 1 dictionary or 10 items")

    #Check to see if one side won
    if units.alive('Att') == 0 and units.alive('Def') == 0:
        return (0,0,1)
    elif units.alive('Att') == 0:
        return (0,1,0)
    elif units.alive('Def') == 0:
        return (1,0,0)

    max_hits = {}
    hits_prob = {}
    # Do we need to copy units to a new object?

    for side in ['Att', 'Def']:
        hits_prob[side] = [0] * (units.alive(side)+1)
        max_hits[side] = units.alive(side)
        for hits in product(*(range(units.units(side)[unit_type] + 1) for unit_type in POSSIBLE_UNITS)):
            total_hits = sum(hits)
            hits_prob[side][total_hits] += (
                prod(binPDF(ODDS[POSSIBLE_UNITS[i]][side] / 6, units.units(side)[POSSIBLE_UNITS[i]], hits[i]) for i in range(len(POSSIBLE_UNITS)))
            )
    prob_win, prob_lose, prob_tie = 0, 0, 0
    odds = 0
    for ahp in range(max_hits['Att'] + 1):
        for dhp in range(max_hits['Def'] + 1):
            if ahp == 0 and dhp == 0:
                continue
            odds = hits_prob['Att'][ahp] * hits_prob['Def'][dhp]
            return_win, return_lose, return_tie = doRound(*casualties(ahp, dhp, units).values())

            # multiply the probabilities of win and loss by the probability of us being here
            prob_win += return_win * odds
            prob_lose += return_lose * odds
            prob_tie += return_tie * odds
    return (prob_win, prob_lose, prob_tie)

def main(units):
    """
    Main function for running the script.
    """
    prob_win, prob_lose, prob_tie = doRound(units)
    total = prob_win + prob_lose + prob_tie
    print(f"{(prob_win / total) * 100:.2f},{(prob_lose / total) * 100:.2f},{(prob_tie / total) * 100:.2f}")

if __name__ == '__main__':
    import sys
    if len(sys.argv) == 1:
        input_units = Units({
            'Att': {'Infantry': 10, 'Tanks': 3},
            'Def': {'Infantry': 10, }
        })
    elif len(sys.argv) == 11:
        input_units = Units([int(arg) for arg in sys.argv[1:]])
    else:
        print('Usage: python aacalc.py [att_inf att_art att_tan att_fig att_bom def_inf def_art def_tan def_fig def_bom]')
        sys.exit(1)
    main(input_units)

    if VERBOSE & 0b01:  # check if the first bit is set
        for func in [
            binPDF, 
            casualties, 
            doRound,
        ]:
            if callable(func) and hasattr(func, 'cache_info'):
                print(func.__name__, "   \t", func.cache_info()) # pylint: disable=no-member disable=no-value-for-parameter
